numpy/_core/src/umath/umathmodule.c

```diff
@@ -99,65 +174,119 @@ ufunc_frompyfunc(PyObject *NPY_UNUSED(dummy), PyObject *args, PyObject *kwds) {
-    if (i) {
-        offset[0] += (sizeof(void *) - i);
-    }
-    offset[1] = nargs;
-    i = (nargs % sizeof(void *));
-    if (i) {
-        offset[1] += (sizeof(void *)-i);
-    }
-    ptr = PyArray_malloc(offset[0] + offset[1] + sizeof(void *) +
-                            (fname_len + 14));
-    if (ptr == NULL) {
+    str = (char *)PyArray_malloc(fname_len + 14);
+    if (str == NULL) {
         Py_XDECREF(pyname);
         return PyErr_NoMemory();
     }
-    fdata = (PyUFunc_PyFuncData *)(ptr);
-    fdata->callable = function;
-    fdata->nin = nin;
-    fdata->nout = nout;
-
-    data = (void **)(((char *)ptr) + offset[0]);
-    data[0] = (void *)fdata;
-    types = (char *)data + sizeof(void *);
-    for (i = 0; i < nargs; i++) {
-        types[i] = NPY_OBJECT;
-    }
-    str = types + offset[1];
     memcpy(str, fname, fname_len);
     memcpy(str+fname_len, " (vectorized)", 14);
     Py_XDECREF(pyname);
 
     /* Do a better job someday */
     doc = "dynamic ufunc based on a python function";
 
+    /* Create the ufunc but without any loops by not passing types*/
     self = (PyUFuncObject *)PyUFunc_FromFuncAndDataAndSignatureAndIdentity(
-            (PyUFuncGenericFunction *)pyfunc_functions, data,
-            types, /* ntypes */ 1, nin, nout, identity ? PyUFunc_IdentityValue : PyUFunc_None,
+            NULL, NULL,
+            NULL, /* ntypes */ 0, nin, nout, identity ? PyUFunc_IdentityValue : PyUFunc_None,
             str, doc, /* unused */ 0, NULL, identity);
 
     if (self == NULL) {
-        PyArray_free(ptr);
+        PyArray_free(str);
+        return NULL;
+    }
+
+    PyArray_DTypeMeta *op_dtypes[NPY_MAXARGS];
+    for (int arg = 0; arg < nargs; arg++) {
+        op_dtypes[arg] = PyArray_DTypeFromTypeNum(NPY_OBJECT);
+        /* These DTypes are immortal and adding INCREFs: so borrow it */
+        Py_DECREF(op_dtypes[arg]);
+    }
+
+    char method_name[101];
+    snprintf(method_name, 100, "ArrayMethod_implementation_for_%s", str);
+
+    NPY_ARRAYMETHOD_FLAGS flags = NPY_METH_REQUIRES_PYAPI | NPY_METH_NO_FLOATINGPOINT_ERRORS;
+
+    PyArrayMethod_GetReductionInitial *get_reduction_intial = NULL;
+    if (nin == 2 && nout == 1) {
+        npy_bool reorderable = NPY_FALSE;
+        PyObject *identity_obj = PyUFunc_GetDefaultIdentity(
+                self, &reorderable);
+        if (identity_obj == NULL) {
+            PyArray_free(str);
+            return NULL;
+        }
+        /*
+         * TODO: For object, "reorderable" is needed(?), because otherwise
+         *       we disable multi-axis reductions `arr.sum(0, 1)`. But for
+         *       `arr = array([["a", "b"], ["c", "d"]], dtype="object")`
+         *       it isn't actually reorderable (order changes result).
+         */
+        if (reorderable) {
+            flags |= NPY_METH_IS_REORDERABLE;
+        }
+        if (identity_obj != Py_None) {
+            get_reduction_intial = &get_initial_from_ufunc;
+        }
+    }
```

Comment:
OK, let's use `get_initial_from_ufunc` for now, although maybe add a `/* TODO: could use a custom get_initial function */` comment?

I do have the ask though to strip down this whole logic to just:
```
/* If the user provided an identity value, assume function is reorderable */
if (identity) {
    flags |= NPY_METH_IS_REORDERABLE;
}
```
if the nin/nout check came from somewhere that seemed important, fine to add it: it doesn't matter in practice anyway.  FWIW, I think it is OK to just set the flag always (even if that suggests that other optimizations might be possible in the future).